File: Insert Delete GetRandom O(1).java
// https://leetcode.com/problems/insert-delete-getrandom-o1/

import java.util.*;

class RandomizedSet {
    
    // Create a hash set
    HashSet<Integer> curSet;

    public RandomizedSet() {
        
        // Allocate memory for hash set
        curSet = new HashSet<Integer>();
    }
    
    public boolean insert(int val) {
        
        // OBJECTIVE: Insert val to hash set in O(1) time
        
        // If val exist, return false
        if (curSet.contains(val)) {return false;}
        
        // Add val to hash set and return true
        curSet.add(val);
        return true;
        
    }
    
    public boolean remove(int val) {
        
        // OBJECTIVE: Remove val from hash set in O(1) time
        
        // If val exist, remove it and return true
        if (curSet.contains(val)) {
            curSet.remove(val);
            return true;
        }
        
        // If val doesn't exist, return false
        return false;
    }
    
    public int getRandom() {
        
        // OBJECTIVE: Return a random element from set
        
        // Get a random index
        int idx = new Random().nextInt(curSet.size());
        
        // Iterate set
        int counter = 0;
        for (int elem: curSet) {
            if (counter == idx) {return elem;}
            counter++;
        }
        
        return 0;
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: Insert Delete GetRandom O(1).py
# https://leetcode.com/problems/insert-delete-getrandom-o1/

from random import choice

class RandomizedSet:

    def __init__(self):
        
        # Create a set
        # NOTE: A set was used because all operations take O(1) time
        self.curSet = set()
        

    def insert(self, val: int) -> bool:
        
        # OBJECTIVE: Insert val in O(1) time
        
        # If element already exist, return true
        if val in self.curSet:
            return False
        
        # Add element to set and return false
        self.curSet.add(val)
        return True

    def remove(self, val: int) -> bool:
        
        # OBJECTIVE: Remove val from set in O(1) time
        
        # If element exist, return true
        if val in self.curSet:
            
            # Create a new set
            self.curSet = self.curSet.difference(set([val]))
            return True
        
        return False

    def getRandom(self) -> int:
        
        # OBJECTIVE: Return a random value from set
        return choice(list(self.curSet))


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
